### PR TITLE
Add LLVM attributes in batches instead of individually

### PR DESCRIPTION
**Pull request by erikdesjardins**
This should improve performance.

~r? @ghost (blocked on #94127)~

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/abi.rs [hunk 1]
```diff
@@ -38,57 +40,68 @@ fn should_use_mutable_noalias(cx: &CodegenCx<'_, '_>) -> bool {
...
-                regular -= ArgAttribute::NonNull;
-            }
-            for (attr, llattr) in OPTIMIZATION_ATTRIBUTES {
-                if regular.contains(attr) {
-                    llattr.apply_llfn(idx, llfn);
-                }
+fn get_attrs<'ll>(
+    this: &ArgAttributes,
+    cx: &CodegenCx<'ll, '_>,
+) -> SmallVec<impl smallvec::Array<Item = &'ll Attribute>> {
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/abi.rs:57]: First time I'm seeing this SmallVec pattern.

File: compiler/rustc_codegen_llvm/src/abi.rs [hunk 2]
```diff
@@ -444,15 +413,14 @@ impl<'ll, 'tcx> FnAbiLlvmExt<'ll, 'tcx> for FnAbi<'tcx, Ty<'tcx>> {
...
-            llvm::Attribute::NoReturn.apply_llfn(llvm::AttributePlace::Function, llfn);
+            func_attrs.push(llvm::AttributeKind::NoReturn.create_attr(cx.llcx));
         }
-
-        // FIXME(eddyb, wesleywiser): apply this to callsites as well?
         if !self.can_unwind {
-            llvm::Attribute::NoUnwind.apply_llfn(llvm::AttributePlace::Function, llfn);
+            func_attrs.push(llvm::AttributeKind::NoUnwind.create_attr(cx.llcx));
         }
+        attributes::apply_to_llfn(llfn, llvm::AttributePlace::Function, &{ func_attrs });
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/abi.rs:423]: Why does this need braces?

**Review by nikic — COMMENTED** [review]
Looks great. As the CI failure indicates, AttributeMask is new in LLVM 14, previously AttrBuilder was used.

File: compiler/rustc_codegen_llvm/src/abi.rs (hunk 2 above)
Comment (erikdesjardins) [reply compiler/rustc_codegen_llvm/src/abi.rs:420]: Not necessary, but causes the SmallVec to be moved, which makes `func_attrs.push()` after this point an error.

**Comment by nikic** [comment]
Doesn't seem to do much for performance, but I think this still makes sense in terms of general design, by separating attribute creation and addition. For example, this allows reusing the same code when applying function and callsite attributes.

@bors r+

**Pull request merged**